spring-data-relational/src/main/java/org/springframework/data/relational/core/mapping/AggregatePath.java

```diff
@@ -0,0 +1,467 @@
+
+		RelationalPersistentEntity<?> leafEntity = getLeafEntity();
+		return leafEntity != null && leafEntity.hasIdProperty();
+	}
+
+	/**
+	 * Returns the longest ancestor path that has an {@link org.springframework.data.annotation.Id} property.
+	 *
+	 * @return A path that starts just as this path but is shorter. Guaranteed to be not {@literal null}.
+	 */
+	public AggregatePath getIdDefiningParentPath() {
+
+		AggregatePath parent = getParentPath();
+
+		if (parent.path == null) {
+			return parent;
+		}
+
+		if (!parent.hasIdProperty()) {
+			return parent.getIdDefiningParentPath();
+		}
+
+		return parent;
+	}
+
+	/**
+	 * The {@link RelationalPersistentEntity} associated with the leaf of this path or throw {@link IllegalStateException}
+	 * if the leaf cannot be resolved.
+	 *
+	 * @return the required {@link RelationalPersistentEntity} associated with the leaf of this path.
+	 * @throws IllegalStateException if the persistent entity cannot be resolved.
+	 */
+	public RelationalPersistentEntity<?> getRequiredLeafEntity() {
+
+		RelationalPersistentEntity<?> entity = getLeafEntity();
+
+		if (entity == null) {
+
+			throw new IllegalStateException(
+					String.format("Couldn't resolve leaf PersistentEntity for type %s", path.getLeafProperty().getActualType()));
+		}
+
+		return entity;
+	}
+
+	public RelationalPersistentProperty getRequiredIdProperty() {
+		return this.path == null ? rootType.getRequiredIdProperty() : getRequiredLeafEntity().getRequiredIdProperty();
+
+	}
+
+	public int getLength() {
+		return path == null ? 0 : path.getLength();
+	}
+
+	/**
+	 * The column name used for the list index or map key of the leaf property of this path.
+	 *
+	 * @return May be {@literal null}.
+	 */
+	@Nullable
+	public SqlIdentifier getQualifierColumn() {
+		return path == null ? null : path.getLeafProperty().getKeyColumn();
+	}
+
+	/**
+	 * The type of the qualifier column of the leaf property of this path or {@literal null} if this is not applicable.
+	 *
+	 * @return may be {@literal null}.
+	 */
+	@Nullable
+	public Class<?> getQualifierColumnType() {
+
+		if (path == null) {
+			return null;
+		}
+		if (!path.getLeafProperty().isQualified()) {
+			return null;
+		}
+		return path.getLeafProperty().getQualifierColumnType();
+	}
+
+	/**
+	 * Returns {@literal true} exactly when the path is non empty and the leaf property an embedded one.
+	 *
+	 * @return if the leaf property is embedded.
+	 */
+	public boolean isEmbedded() {
+		return path != null && path.getLeafProperty().isEmbedded();
+	}
+
+	/**
+	 * @return {@literal true} when this is an empty path or the path references an entity.
+	 */
+	public boolean isEntity() {
+		return path == null || path.getLeafProperty().isEntity();
+	}
+
+	/**
+	 * Finds and returns the longest path with ich identical or an ancestor to the current path and maps directly to a
+	 * table.
+	 *
+	 * @return a path. Guaranteed to be not {@literal null}.
+	 */
+	private AggregatePath getTableOwningAncestor() {
```

Comment:
This method is another indication of the need to traverse the `AggregatePath` hierarchy with having an utility for querying the model.